app/Jobs/TweetNewIssue.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace App\Jobs;
+
+use App\DataTransferObjects\Issue;
+use App\Facades\Twitter;
+use App\Models\SocialPost;
+use Cache;
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldBeUnique;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Foundation\Bus\Dispatchable;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Queue\Middleware\RateLimited;
+use Illuminate\Queue\SerializesModels;
+use RateLimiter;
+
+class TweetNewIssue implements ShouldQueue, ShouldBeUnique
+{
+    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;
+
+    public function __construct(protected Issue $issue)
+    {
+        //
+    }
+
+    public function handle(): void
+    {
+        Cache::lock('twitter-lock')->block(10, function () {
+            $this->tweetAboutIssue();
+        });
+    }
+
+    protected function tweetAboutIssue(): void
+    {
+        $socialPost = SocialPost::firstOrNew(
+            ['issue_id' => $this->issue->id],
+            [
+                'issue_repo' => $this->issue->repoName,
+                'issue_number' => $this->issue->number,
+            ]
+        );
+
+        if ($socialPost->tweetWasSent()) {
+            return;
+        }
+
+        $response = Twitter::tweet("An issue in {$this->issue->repoName} may need your help: {$this->issue->title}" . PHP_EOL . $this->issue->url);
+
+        if (empty($response['data']->id)) {
+            RateLimiter::clear('twitter');
+            return;
+        }
+
+        $socialPost->fill([
+            'twitter_sent_at' => now(),
+            'tweet_id' => $response['data']->id,
+        ]);
+
+        $socialPost->save();
+    }
+
+    /**
+     * @return array<int, object>
+     */
+    public function middleware(): array
+    {
+        return [new RateLimited('twitter')];
```

Comment:
I might be wrong, so feel free to correct me if so. But, I think if we use this middleware then the job will be put back on the queue if the rate limit is exceeded.

So if we have 10 tweets to send, the first will be sent and then the other 9 will be pushed back onto the queue to retry later.

I'm wondering if we should update this line to

```php
return [(new RateLimited('twitter'))->dontRelease()];
```

So after we've sent one tweet in an hour, we discard all the other tweets? And then they'll be picked up on the next run-through in an hour?

If I'm wrong on this though, my apologies 🙂

Reply:
That would work slightly better, yes. I have changed it!